### PR TITLE
Add ability for slices to listen to other actions

### PR DESCRIPTION
**Pull request by markerikson**
Thus far, slices generated by `createSlice` could only listen to the specific action types defined for each field in the `reducers` option.

I've added a field named `extraReducers` that lets you add reducers that listen to action types that were defined elsewhere.

I'm open to bikeshedding suggestions on the field name.  Other ideas:

```
otherReducers
listenForActions
handleActions
```

I also updated `createAction` so that action creators have the string available as a `type` field. This was requested in #72 for use with switch statements.  Turns out it also helps with TS usage, because TS won't let you pass a function directly as a computed object property - it's not a string or `any`.  So, `[actionCreator.type]` is a passable alternative.

The TS compiler _seems_ to be okay with this at the moment, but this is the first time I've ever done anything meaningful with TS, so I'd appreciate review.  (Paging @denisw , @Dudeonyx , and @jessidhia ...) 

I filled out the `createSlice` docs as well.

### CODE REVIEW

File: docs/api/createSlice.md [hunk 1]
```diff
@@ -43,18 +147,18 @@ const reducer = combineReducers({
 const store = createStore(reducer)
 
 store.dispatch(counter.actions.increment())
-// -> { counter: 1, user: {} }
+// -> { counter: 1, user: {name : '', age: 20} }
```
Comment (BTMPL) [inline docs/api/createSlice.md]: The `age` should read 21 in this case.

**Comment by Dudeonyx** [comment]
@markerikson instead of creating a new field called `extraReducers` to enable the reducer listen to other actions it would be far easier to change this in `createSlice.ts` `line: 102`

```ts
  const actionMap = actionKeys.reduce(
    (map, action) => {
      const type = getType(slice, action)
      map[action] = createAction(type)
      return map
    },
    {} as any
  )
```

to

```ts
  const actionMap = actionKeys.reduce(
    (map, action) => {
      map[action] = createAction(action)  // <- this specifically
      return map
    },
    {} as any
  )
```

By removing the `getType()` function which attaches the `slice` to the action type, the `actionCreators` `type` strings are no longer unique to each reducer, much like a regular switch-case reducer.

So instead of 

```ts
const user = createSlice({
  slice: 'user',
  initialState: { name: '', age: 20 },
  reducers: {
    setUserName: (state, action) => {
      state.name = action.payload // mutate the state all you want with immer
    }
  },
  extraReducers: {
    [counter.actions.increment]: (state, action) => {
      state.age += 1
    }
  }
})
```

we do

```ts
const user = createSlice({
  slice: 'user',
  initialState: { name: '', age: 20 },
  reducers: {
    setUserName: (state, action) => {
      state.name = action.payload // mutate the state all you want with immer
    },
   [counter.actions.increment]: (state, action) => {
      state.age += 1
    }
  },
})
```

**Comment by markerikson** [comment]
@Dudeonyx : maybe I'm misreading the changes you're suggesting, but it seems like they break part of the point of what `createSlice` is intended to do.

I specifically _want_ `createSlice` to generated prefixed action types if a slice name is provided.  I also specifically want to distinguish between "these are action types that are 'owned' by this slice, generate action creators for them" and "these are action types from somewhere else, _don't_ generate action creators".

File: type-tests/files/createSlice.typetest.ts [hunk 2]
```diff
@@ -15,6 +15,9 @@ import {
     reducers: {
       increment: (state: number, action) => state + action.payload,
       decrement: (state: number, action) => state - action.payload
+    },
+    extraReducers: {
+      "OTHER_ACTION_TYPE" : (state : number, action ) => state + action.payload.count
```
[the remark this answers is not in the document]
Comment (markerikson) [reply type-tests/files/createSlice.typetest.ts]: No, I don't have Prettier configured in my editor at all.  Probably something I should set up.

**Comment by markerikson** [comment]
Closing this PR in favor of #86 , since that's a superset.

I just opened up #91 to discuss how `createSlice` ought to behave beyond this capability.

**Comment by markerikson** [comment]
Didn't realize #86 was really meant for this branch.  Merged it over, and putting this in.

**Pull request merged**